packages/common/nbstore/src/sync/blob/index.ts

```diff
@@ -1,56 +1,84 @@
+import EventEmitter2 from 'eventemitter2';
 import { difference } from 'lodash-es';
+import { BehaviorSubject, type Observable } from 'rxjs';
 
 import type { BlobRecord, BlobStorage } from '../../storage';
+import { OverCapacityError } from '../../storage';
 import { MANUALLY_STOP, throwIfAborted } from '../../utils/throw-if-aborted';
+import type { PeerStorageOptions } from '../types';
+
+export interface BlobSyncState {
+  isStorageOverCapacity: boolean;
+}
 
 export interface BlobSync {
+  readonly state$: Observable<BlobSyncState>;
   downloadBlob(
     blobId: string,
     signal?: AbortSignal
   ): Promise<BlobRecord | null>;
   uploadBlob(blob: BlobRecord, signal?: AbortSignal): Promise<void>;
+  fullSync(signal?: AbortSignal): Promise<void>;
+  setMaxBlobSize(size: number): void;
+  onReachedMaxBlobSize(cb: (byteSize: number) => void): () => void;
 }
 
 export class BlobSyncImpl implements BlobSync {
+  readonly state$ = new BehaviorSubject<BlobSyncState>({
+    isStorageOverCapacity: false,
+  });
   private abort: AbortController | null = null;
+  private maxBlobSize: number = 1024 * 1024 * 100; // 100MB
```

Comment:
it's dynamic, and the highest possible we promised to user is 500MB

Reply:
This value will be set on the frontend.

The ideal implementation here would be to try calling `storage.set` and then limits the size based on the error returned from the storage.

But the current implementation is just copied from the old blobStorage.